rust/src/pet/participant.rs

```diff
@@ -1,122 +1,306 @@
+            Ok(())
+        } else if is_eligible(&self.signature_update, round_update)
+            .ok_or(PetError::InvalidMessage)?
+        {
+            self.task = Task::Update;
+            Ok(())
+        } else {
+            self.task = Task::None;
+            Ok(())
+        }
+    }
 }
 
-impl Message {
-    pub fn new(round_sum: f64, round_update: f64, round_seed: Vec<u8>) -> Self {
+impl Default for Participant {
+    fn default() -> Self {
+        let (encr_pk, encr_sk) = box_::gen_keypair();
+        let (sign_pk, sign_sk) = sign::gen_keypair();
+        let ephm_pk = box_::PublicKey([0_u8; box_::PUBLICKEYBYTES]);
+        let ephm_sk = box_::SecretKey([0_u8; box_::SECRETKEYBYTES]);
+        let certificate: Vec<u8> = Vec::new();
+        let signature_sum = sign::Signature([0_u8; sign::SIGNATUREBYTES]);
+        let signature_update = sign::Signature([0_u8; sign::SIGNATUREBYTES]);
+        let task = Task::None;
         Self {
-            round_sum,
-            round_update,
-            round_seed,
+            encr_pk,
+            encr_sk,
+            sign_pk,
+            sign_sk,
+            ephm_pk,
+            ephm_sk,
+            certificate,
+            signature_sum,
+            signature_update,
+            task,
         }
     }
+}
+
+// Message egress with buffers:
+//
+// encr_pk -┐
+// sign_pk -┤
+//          └-> SealedBoxBuffer
+//               └-> SealedBox -------┐
+// certificate ------┐                |
+// signature_sum ----┤                |
+// signature_update -┤                |
+// ephm_pk ----------┤                |
+//                   └-> SumBoxBuffer |
+//                        └-> SumBox -┤
+//                                    └-> MessageBuffer
+//                                         └-> SumMessage
+//
+// encr_pk -┐
+// sign_pk -┤
+//          └-> SealedBoxBuffer
+//               └-> SealedBox ----------┐
+// certificate ------┐                   |
+// signature_sum ----┤                   |
+// signature_update -┤                   |
+// model_url---------┤                   |
+// dict_seed---------┤                   |
+//                   └-> UpdateBoxBuffer |
+//                        └-> UpdateBox -┤
+//                                       └-> MessageBuffer
+//                                            └-> UpdateMessage
+//
+// encr_pk -┐
+// sign_pk -┤
+//          └-> SealedBoxBuffer
+//               └-> SealedBox --------┐
+// certificate ------┐                 |
+// signature_sum ----┤                 |
+// signature_update -┤                 |
+// mask_url ---------┤                 |
+//                   └-> Sum2BoxBuffer |
+//                        └-> Sum2Box -┤
+//                                     └-> MessageBuffer
+//                                          └-> Sum2Message
+
+/// Buffer and wrap the asymmetrically encrypted part of a "sum/update/sum2" message.
+struct SealedBoxBuffer<'a>(&'a [u8], &'a [u8], &'a [u8]);
```

Comment:
This works, but I think it would be slightly more user friendly to not require the lifetimes to be the same, to make things a bit more flexible. Also giving names to the lifetimes sometimes helps the reader to better understand what's going on.

```
struct SealedBoxBuffer<'sum, 'update, 'sum2>(&'sum [u8], &'update [u8], &'sum2 [u8]);
```

Reply:
i'll add lifetime naming, it definitely helps readability and adds flexibility. i would keep the tuple struct and not change it to to a struct, since it is a private object without an external api anyways.